compiler/rustc_ty_utils/src/implied_bounds.rs

```diff
@@ -56,6 +67,38 @@ fn assumed_wf_types<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> &'tcx ty::List<Ty
         | DefKind::LifetimeParam
         | DefKind::GlobalAsm
         | DefKind::Closure
-        | DefKind::Generator => ty::List::empty(),
+        | DefKind::Generator => ty::EarlyBinder(ty::Binder::dummy(ty::List::empty())),
     }
 }
+
+fn from_fn_sig<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+) -> ty::Binder<'tcx, &'tcx ty::List<Ty<'tcx>>> {
+    // FIXME(#84533): We probably shouldn't use output types in implied bounds.
+    // This would reject this fn `fn f<'a, 'b>() -> &'a &'b () { .. }`.
```

Comment:
I don't think this should be how we fix #84533 because you'd still have an builtin impl with an associated type which isn't well formed.

The correct fix is to actually check that the return type is well formed even without calling the function. This needs implications for binders though.

Until then something like #106807 might work as a stopgap (again having the issue of unnormalized projections though)